stor/dx.py

```diff
@@ -144,41 +138,273 @@ def download_objects(self):  # may not need it
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        new_file_h = file_handler.clone(project=dest.canonical_project,
+                                        folder=(folder_dest.resource or '/'))
+        if not dest_exists and not dest.endswith('/'):
+            new_file_h.rename(dest.name)
+
+    def _move(self, dest):
+        """Moves the data object to a different folder within project
+
+        Args:
+            dest (path) : The destination file/folder path within same project
+
+        Raises:
+            ValueError : When attempting to move projects
+            DNAnexusError : If attempting to move across projects
+        """
+        if not self.resource:
+            raise ValueError('Cannot move project ({})'.format(self))
+        if dest.canonical_project != self.canonical_project:
+            # This can be implemented by clone and remove original
+            raise DNAnexusError('Cannot move across different projects')
+        file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                   project=self.canonical_project)
+        dest_exists = dest.isdir()
+        if dest_exists:
+            new_dest = dest / self.name
+            if new_dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(new_dest))
+            folder_dest = dest
+        else:
+            if dest.isfile():
+                raise DuplicateError('Dest path ({}) already exists. Do not duplicate!'
+                                     .format(dest))
+            folder_dest = dest.parent
+            folder_dest.makedirs_p()
+
+        file_handler.move(folder_dest.resource or '/')
+
+        if not dest_exists and not dest.endswith('/'):
+            file_handler.rename(dest.name)
+
+    def copy(self, dest, move_within_project=False):
+        """Copies data object to destination path.
+
+        Args:
+            dest (path|str): The destination file or directory.
+            move_within_project (bool): If True, make new files in DX instead of cloning.
+                Only comes into effect when both source and destination are DXPaths
+
+        Raises:
+            DNAnexusError : When cloning within same project with copy_new=False
+            DuplicateError : When cloning data object or folder to already existing path
+        """
+        dest = Path(dest)
+        if utils.is_dx_path(dest):
+            if self.isfile():
+                if dest.canonical_project == self.canonical_project:
+                    if move_within_project:
+                        self._move(dest)
+                    else:
+                        raise DNAnexusError('Cannot copy within same project without flag set')
```

Comment:
Let's make this a clearer / more specific error class.

Additionally, phrasing probably better as `Source and destination are in same project. Set move_within_project=True to allow this.`

Also I think this might be better as:

```
try:
   self._clone(dest)
except AlreadyInProjectError:
   if not move_within_project: ...
```

because the file could already be in the project even though the src and dest are not in the same project, e.g. this test case:

```
main_p = Path('dx://P:/a/mypath.txt')
with main_p.open('w') as fp:
    fp.write('blah')
proj2 = Path('dx://P2')
path2 = Path('dx://P2:/mynewpath.txt')
main_p.copy(path2)
# raise?
path2.copy('dx://P:/someotherpath.txt', move_within_project=False)
path2.copy('dx://P:/someotherpath.txt', move_within_project=True)
```

Reply:
I agree, I wanted to do it as `try:
   self._clone(dest)
except AlreadyInProjectError:
   if not move_within_project: ...` too. The only thing is _clone calls dxpy's file_handler.clone or project_handler.clone, and 'shockingly', dxpy doesn't error if the file is already in the destination project. It doesn't even move the file or anything. It in fact completely ignores the command, and you won't know for sure if your file is actually cloned into the destination unless you go manually check by `ls`.

I did change the paraphrasing though.

Reply:
Also, I created DNAnexusError for stuff that is disallowed from DX's side, just for your reference. I think this fits right there. Let me know.